File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_typescript-vite/src/lib/handlers/npcsFactions.ts
import type { ServerResponse } from 'node:http';
import { sendError, sendJson } from '../http.js';
import {
  campaignExists,
  countFactionsByCampaign,
  countFriendlyNpcsByCampaign,
  countNpcsByCampaign,
  createFaction,
  createNpc,
  factionExists,
  getFaction,
  npcExists,
} from '../db.js';
import { isDisposition, isNonEmptyString, isStance } from '../validation.js';
import type { Faction, Npc } from '../types.js';

export function handleCreateFaction(pathname: string, body: unknown, res: ServerResponse): boolean {
  const match = pathname.match(/^\/v1\/campaigns\/(.+)\/factions$/);
  if (!match) return false;
  const campaignId = match[1];
  if (!campaignExists(campaignId)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  const b = body as any;
  if (!b || !isNonEmptyString(b.id) || !isNonEmptyString(b.name) || !isStance(b.stance)) {
    sendError(res, 400, 'invalid request');
    return true;
  }
  if (factionExists(b.id)) {
    sendError(res, 409, 'faction already exists');
    return true;
  }
  const faction: Faction = {
    id: b.id,
    campaign_id: campaignId,
    name: b.name,
    stance: b.stance,
  };
  createFaction(faction);
  sendJson(res, 201, { id: faction.id, name: faction.name, stance: faction.stance });
  return true;
}

export function handleCreateNpc(pathname: string, body: unknown, res: ServerResponse): boolean {
  const match = pathname.match(/^\/v1\/campaigns\/(.+)\/npcs$/);
  if (!match) return false;
  const campaignId = match[1];
  if (!campaignExists(campaignId)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  const b = body as any;
  if (!b || !isNonEmptyString(b.id) || !isNonEmptyString(b.name) || !isNonEmptyString(b.faction_id) || !isDisposition(b.disposition)) {
    sendError(res, 400, 'invalid request');
    return true;
  }
  if (npcExists(b.id)) {
    sendError(res, 409, 'npc already exists');
    return true;
  }
  const faction = getFaction(b.faction_id);
  if (!faction || faction.campaign_id !== campaignId) {
    sendError(res, 400, 'faction not found');
    return true;
  }
  const npc: Npc = {
    id: b.id,
    campaign_id: campaignId,
    name: b.name,
    faction_id: b.faction_id,
    disposition: b.disposition,
  };
  createNpc(npc);
  sendJson(res, 201, {
    id: npc.id,
    name: npc.name,
    faction_id: npc.faction_id,
    disposition: npc.disposition,
  });
  return true;
}

export function handleGetRelationships(pathname: string, res: ServerResponse): boolean {
  const match = pathname.match(/^\/v1\/campaigns\/(.+)\/relationships$/);
  if (!match) return false;
  const campaignId = match[1];
  if (!campaignExists(campaignId)) {
    sendError(res, 404, 'campaign not found');
    return true;
  }
  sendJson(res, 200, {
    campaign_id: campaignId,
    factions: countFactionsByCampaign(campaignId),
    npcs: countNpcsByCampaign(campaignId),
    friendly_npcs: countFriendlyNpcsByCampaign(campaignId),
  });
  return true;
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T052954Z_claude_claude-sonnet-5_python-flask/dndapp/routes/npcs_factions.py
"""Campaign factions, NPCs, and their aggregate relationship state."""

from flask import Blueprint, jsonify, request

from ..db import get_db
from ..validation import valid_int, valid_nonempty_str

bp = Blueprint("npcs_factions", __name__)


@bp.post("/v1/campaigns/<campaign_id>/factions")
def create_faction(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        data = request.get_json(silent=True) or {}
        faction_id = data.get("id")
        name = data.get("name")
        stance = data.get("stance")

        if not isinstance(faction_id, str) or not faction_id:
            return jsonify(error="invalid id"), 400
        if not valid_nonempty_str(name):
            return jsonify(error="invalid name"), 400
        if not valid_nonempty_str(stance):
            return jsonify(error="invalid stance"), 400

        existing = conn.execute(
            "SELECT id FROM factions WHERE campaign_id = ? AND id = ?",
            (campaign_id, faction_id),
        ).fetchone()
        if existing is not None:
            return jsonify(error="faction already exists"), 409

        conn.execute(
            "INSERT INTO factions (campaign_id, id, name, stance) VALUES (?, ?, ?, ?)",
            (campaign_id, faction_id, name, stance),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(id=faction_id, name=name, stance=stance), 201


@bp.post("/v1/campaigns/<campaign_id>/npcs")
def create_npc(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        data = request.get_json(silent=True) or {}
        npc_id = data.get("id")
        name = data.get("name")
        faction_id = data.get("faction_id")
        disposition = data.get("disposition")

        if not isinstance(npc_id, str) or not npc_id:
            return jsonify(error="invalid id"), 400
        if not valid_nonempty_str(name):
            return jsonify(error="invalid name"), 400
        if faction_id is not None and not isinstance(faction_id, str):
            return jsonify(error="invalid faction_id"), 400
        if not valid_int(disposition):
            return jsonify(error="invalid disposition"), 400

        if faction_id is not None:
            faction = conn.execute(
                "SELECT id FROM factions WHERE campaign_id = ? AND id = ?",
                (campaign_id, faction_id),
            ).fetchone()
            if faction is None:
                return jsonify(error="faction not found"), 404

        existing = conn.execute(
            "SELECT id FROM npcs WHERE campaign_id = ? AND id = ?",
            (campaign_id, npc_id),
        ).fetchone()
        if existing is not None:
            return jsonify(error="npc already exists"), 409

        conn.execute(
            "INSERT INTO npcs (campaign_id, id, name, faction_id, disposition) "
            "VALUES (?, ?, ?, ?, ?)",
            (campaign_id, npc_id, name, faction_id, disposition),
        )
        conn.commit()
    finally:
        conn.close()

    return jsonify(
        id=npc_id, name=name, faction_id=faction_id, disposition=disposition
    ), 201


@bp.get("/v1/campaigns/<campaign_id>/relationships")
def get_relationships(campaign_id):
    conn = get_db()
    try:
        campaign = conn.execute(
            "SELECT id FROM campaigns WHERE id = ?", (campaign_id,)
        ).fetchone()
        if campaign is None:
            return jsonify(error="campaign not found"), 404

        faction_count = conn.execute(
            "SELECT COUNT(*) AS c FROM factions WHERE campaign_id = ?",
            (campaign_id,),
        ).fetchone()["c"]
        npc_count = conn.execute(
            "SELECT COUNT(*) AS c FROM npcs WHERE campaign_id = ?",
            (campaign_id,),
        ).fetchone()["c"]
        friendly_npc_count = conn.execute(
            "SELECT COUNT(*) AS c FROM npcs n "
            "JOIN factions f ON f.campaign_id = n.campaign_id AND f.id = n.faction_id "
            "WHERE n.campaign_id = ? AND f.stance = 'friendly'",
            (campaign_id,),
        ).fetchone()["c"]
    finally:
        conn.close()

    return jsonify(
        campaign_id=campaign_id,
        factions=faction_count,
        npcs=npc_count,
        friendly_npcs=friendly_npc_count,
    )
